runtime/sema/checker.go

```diff
@@ -2069,37 +2069,74 @@ func (checker *Checker) checkInvalidInterfaceAsType(ty Type, pos ast.HasPosition
 }
 
 func (checker *Checker) checkParameterizedTypeIsInstantiated(ty Type, pos ast.HasPosition) {
-	parameterizedType, ok := ty.(ParameterizedType)
-	if !ok {
-		return
-	}
+	switch t := ty.(type) {
+	case *OptionalType:
+		checker.checkParameterizedTypeIsInstantiated(t.Type, pos)
 
-	typeArgs := parameterizedType.TypeArguments()
-	typeParameters := parameterizedType.TypeParameters()
+	case ArrayType:
+		checker.checkParameterizedTypeIsInstantiated(t.ElementType(false), pos)
 
-	typeArgumentCount := len(typeArgs)
-	typeParameterCount := len(typeParameters)
+	case *DictionaryType:
+		checker.checkParameterizedTypeIsInstantiated(t.KeyType, pos)
+		checker.checkParameterizedTypeIsInstantiated(t.ValueType, pos)
 
-	if typeArgumentCount != typeParameterCount {
-		checker.report(
-			&InvalidTypeArgumentCountError{
-				TypeParameterCount: typeParameterCount,
-				TypeArgumentCount:  typeArgumentCount,
-				Range: ast.NewRange(
-					checker.memoryGauge,
-					pos.StartPosition(),
-					pos.EndPosition(checker.memoryGauge),
-				),
-			},
-		)
-	}
+	case *ReferenceType:
+		checker.checkParameterizedTypeIsInstantiated(t.Type, pos)
+
+	case *FunctionType:
+		for _, tyParam := range t.TypeParameters {
+			checker.checkParameterizedTypeIsInstantiated(tyParam.TypeBound, pos)
+		}
+
+		for _, param := range t.Parameters {
+			checker.checkParameterizedTypeIsInstantiated(param.TypeAnnotation.Type, pos)
+		}
+
+		checker.checkParameterizedTypeIsInstantiated(t.ReturnTypeAnnotation.Type, pos)
 
-	// Ensure that each non-optional typeparameter is non-nil.
-	for index, typeParam := range typeParameters {
-		if !typeParam.Optional && typeArgs[index] == nil {
+	case *RestrictedType:
+		checker.checkParameterizedTypeIsInstantiated(t.Type, pos)
+
+	case *CompositeType:
+		if t.EnumRawType != nil {
+			checker.checkParameterizedTypeIsInstantiated(t.EnumRawType, pos)
+		}
+		if t.containerType != nil {
+			checker.checkParameterizedTypeIsInstantiated(t.containerType, pos)
+		}
```

Comment:
`containerType` here and below is the type inside of which this composite type is declared, e.g. in a nested declaration `contract C { struct S {} }`, `C` is the container type of `S`.

The current implementation recursed "upwards", so this will lead to an infinite recursion. Can you please add tests for nested types, for composite types and for interface types below?

Reply:
Done.